Weekly Mood Insights API

buyekeobare opened the pull request:
**Description:**
Implements the /api/v1/insights/history endpoint 

**Features:**

**- GET /api/v1/insights/history?week_offset=N → returns:**
- [ ]  graph: 7 mood points (Mon–Sun)
- [ ] - timeline: per-day note snippets
**- Supports:**
- [ ] Missing days (graph shows null)
- [ ] Mood snapshot logic (mood_value_snapshot preferred)
- [ ] Efficient query using latest note per day
- [ ] Caching per (user, week_offset) for 5 minutes

**Backend Implementation:**

- insights/views.py: APIView + request validation
- insights/services.py: week range logic, cache, timeline, graph
- notes/signals.py: auto-fill mood_value_snapshot from Mood.category
- notes/migrations/0005_backfill_mood_value_snapshot.py: backfills existing notes
- insights/tests/test_history.py: full test suite (7d graph, user scoping, nulls, cache, snapshot)
- insights/urls.py: added named route insights-history

 **Logic:**

- insights/services.py builds payload using fetch_weekly_latest_per_day() 
- insights/views.py handles request validation
- notes/signals.py auto-fills snapshot via pre_save
- Fallback: maps Mood.category (string) to 1–5 scale
- Caching via Django’s cache framework
- Snapshot backfill migration included

**Checklist:**

- GET /api/v1/insights/history returns graph + timeline
-  Accepts week_offset or week as query param
-  Uses Note.mood_value_snapshot or fallback category
-  Signal auto-fills snapshot on note create/update
-  Backfill migration for historical notes
-  Unit tests for graph, timeline, cache, scoping, null days
-  Caching with TTL = 300s
-  API documented via example and internal docstring

**How to Test:**

GET /api/v1/insights/history?week_offset=0
Authorization: Token <user-token>

**Expect Response:**
<img width="829" height="459" alt="Insights Response" src="https://github.com/user-attachments/assets/27c88c7f-77f7-405d-a134-0da7b8fdc145" />


- 7 graph points (some may be null)
- timeline with snippets, mood_id, and date

**Other Changes:**

Updated INSTALLED_APPS:
"notes.apps.NotesConfig"

→ Enables Django to load signals.py, ensuring Note.mood_value_snapshot is auto-filled correctly.

**Recent Updates**

**File Updated: `backend/insights/services.py`**

- **Clarified time zone logic** in `get_week_range()` docstring to explicitly state that it uses the active time zone as defined by Django's `TIME_ZONE` setting.
- **Optimized mood query** in `fetch_weekly_latest_per_day()` by adding `select_related("mood")` to eliminate N+1 DB queries.
- **Adjusted `week.end` value** in the `/insights/history` response to be inclusive (i.e., returns Sunday as the final day), improving frontend alignment.
- **Removed duplicated comment header** for cleaner code readability.

**Test Coverage Added**

- Verified `week.end` is inclusive (Sunday) using date assertions.
- Confirmed support for the `?week=` alias parameter (same behavior as `week_offset`).
- Asserted that every graph point includes a `mood_id` field (even if `null`).
- Ensured `mood_value_snapshot` overrides any mood.category-derived fallback.
- Added test to check snippet truncation at 120 characters with ellipsis.

These additions improve confidence in the `/api/v1/insights/history` contract and ensure edge cases are fully covered.

**Related Issue**

Closes #73

buyekeobare committed:
feat(insights): add week range & mood aggregation logic for history endpoint

buyekeobare committed:
feat(insights): add InsightsHistoryView for weekly graph and timeline (week_offset support)

buyekeobare committed:
chore(insights): clean up and route InsightsHistoryView under /api/v1/insights/history

buyekeobare committed:
test(insights): add robust versioned tests for InsightsHistoryView (graph, timeline, validation, cache)

buyekeobare committed:
chore(insights): route history/tip under versioned prefix via core urls

buyekeobare committed:
feat(notes): add mood_value_snapshot for insights rollups

buyekeobare committed:
fix(insights): use Note.note for timeline snippet; honor mood_value_snapshot when present

buyekeobare committed:
test(insights): fix cache test tail; ensure clean suite

buyekeobare committed:
fix(insights): map Mood.category string to 1..5 in _mood_value_for; prefer snapshot when present

buyekeobare committed:
feat(notes): auto-fill mood_value_snapshot via pre_save signal; sync on mood changes

buyekeobare committed:
test(notes): verify snapshot auto-fill on create and mood change; preserve explicit values

buyekeobare committed:
chore(settings): register NotesConfig to enable signal loading in notes app

EgbaFrank reviewed (commented):
Good work, a robust implementation.

buyekeobare committed:
perf(insights-history): optimize note query with select_related and return inclusive week end date

buyekeobare commented on backend/insights/services.py:
```diff
@@ -0,0 +1,151 @@
+# insights/services.py
+# insights/services.py
+from dataclasses import dataclass
```
Resolved: Removed the duplicate comment header

buyekeobare commented:
Update: Added a small optimization with select_related("mood") to avoid N+1 queries and adjusted week.end in the response to be inclusive (Sunday). No breaking changes.

buyekeobare committed:
docs(insights): clarify timezone behavior in get_week_range docstring

buyekeobare commented on backend/insights/services.py:
```diff
@@ -0,0 +1,151 @@
+    """
+    Monday-based weeks in the project's TIME_ZONE.
+    week_offset=0 -> current week; 1 -> previous week; etc.
+    """
```
Resolved: Updated the docstring to clarify that the week range uses the active time zone as defined by Django’s TIME_ZONE setting. Thanks for the suggestion!

buyekeobare committed:
test(insights-history): add coverage for inclusive week range, alias param, mood_id, snapshot override, and snippet truncation

buyekeobare committed:
docs(insights): align history response shape (nested week, mood_id), clarify validation differences vs tip

buyekeobare committed:
revert(frontend): remove unintended frontend changes from this branch

buyekeobare committed:
revert(frontend): reset App.jsx to HEAD~3 state

EgbaFrank reviewed (commented):
Good work
